geometry/cartesian_product.hpp

```diff
@@ -37,11 +37,23 @@ struct CartesianProductAdditiveGroup;
 template<typename Scalar,
          typename Tuple,
          typename = std::make_integer_sequence<int, std::tuple_size_v<Tuple>>>
-struct CartesianProductVectorSpace;
+class CartesianProductVectorSpace;
+
+// This class exports one function:
+//
+//   static constexpr ... Multiply(LTuple const& left, RTuple const& right);
+//
+// where the return types are appropriate for the operation.  The element types
+// of the tuples must have a structure of multiplicative ring.
+template<typename LTuple, typename RTuple,
+         int lsize_ = std::tuple_size_v<LTuple>,
+         int rsize_ = std::tuple_size_v<RTuple>>
+class CartesianProductRing;
```

Comment:
CartesianProductAdditiveGroup and CartesianProductVectorSpace make sense as names (they are the direct product of groups and direct product of vector spaces respectively), but this has to be called PolynomialRing; it is emphatically *not* the product ring.

Reply:
Done.